zebra-rpc/src/queue.rs

```diff
@@ -0,0 +1,322 @@
+        self.sender.clone()
+    }
+
+    /// Create a new receiver.
+    pub fn receiver(&self) -> Receiver<Option<UnminedTx>> {
+        self.sender.subscribe()
+    }
+
+    /// Get the queue transactions as a `HashSet` of unmined ids.
+    fn transactions_as_hash_set(&self) -> HashSet<UnminedTxId> {
+        let transactions = self.queue.transactions();
+        transactions.iter().map(|t| *t.0).collect()
+    }
+
+    /// Get the queue transactions as a `Vec` of transactions.
+    fn transactions_as_vec(&self) -> Vec<Arc<Transaction>> {
+        let transactions = self.queue.transactions();
+        transactions.iter().map(|t| t.1 .0.clone()).collect()
+    }
+
+    /// Update the `tip_height` field with a new height.
+    pub fn update_tip_height(&mut self, height: Height) {
+        self.tip_height = height;
+    }
+
+    /// Retry sending to memempool if needed.
+    ///
+    /// Creates a loop that will run each time a new block is mined.
+    /// In this loop, get the transactions that are in the queue and:
+    /// - Check if they are now in the mempool and if so, delete the transaction from the queue.
+    /// - Check if the transaction is now part of a block in the state and if so,
+    /// delete the transaction from the queue.
+    /// - With the transactions left in the queue, retry sending them to the mempool ignoring
+    /// the result of this operation.
+    ///
+    /// Addtionally, each iteration of the above loop, will receive and insert to the queue
+    /// transactions that are pending in the channel.
+    pub async fn run<Mempool, State, Tip>(
+        mut self,
+        mempool: Mempool,
+        state: State,
+        tip: Tip,
+        network: Network,
+    ) where
+        Mempool: Service<Request, Response = Response, Error = BoxError> + Clone + 'static,
+        State: Service<ReadRequest, Response = ReadResponse, Error = zebra_state::BoxError>
+            + Clone
+            + Send
+            + Sync
+            + 'static,
+        Tip: ChainTip + Clone + Send + Sync + 'static,
+    {
+        let mut receiver = self.sender.subscribe();
+
+        loop {
+            // if we don't have a chain use `NO_CHAIN_TIP_HEIGHT` to get block spacing
+            let tip_height = match tip.best_tip_height() {
+                Some(height) => height,
+                _ => NO_CHAIN_TIP_HEIGHT,
+            };
+
+            // get spacing between blocks
+            let spacing = NetworkUpgrade::target_spacing_for_height(network, tip_height);
+
+            // sleep until the next block
+            tokio::time::sleep(spacing.to_std().expect("should never be less than zero")).await;
+
+            // get transactions from the channel
+            while let Ok(Some(tx)) = receiver.try_recv() {
+                let _ = &self.queue.insert(tx.clone());
+            }
+
+            // skip some work if stored tip height is the same as the one arriving
+            if tip_height > self.tip_height {
```

Comment:
Minor: I forgot about chain reorganisations here, so I gave you some bad advice.

If Zebra switches between chain forks, the height can be the same or lower, but the tip block will be different. To handle this case, we could store the tip block hash instead of the height. (Block hashes are unique.)

This code will still work most of the time. Only 1 in 300 blocks have a fork.

So we could do this quick fix instead:
```suggestion
            // TODO: check tip block hashes instead, so we always retry when there is a chain fork (these are rare)
            if tip_height != self.tip_height {
```

Reply:
https://github.com/ZcashFoundation/zebra/pull/4015/commits/4e3229bf3f1f972fa7fbc90170fe639ad0ba65d8